pallets/verifiers/src/tests.rs

```diff
@@ -95,32 +137,164 @@ mod register_should {
         assert_eq!(info.pays_fee, Pays::Yes);
         assert_eq!(info.weight, MockWeightInfo::register_vk(&43));
     }
-}
 
-pub mod submit_proof_should {
-    use hex_literal::hex;
+    #[rstest]
+    fn hold_a_deposit(mut test_ext: sp_io::TestExternalities) {
+        test_ext.execute_with(|| {
+            let initial_reserved_balance = Balances::reserved_balance(USER_1);
+            assert_ok!(FakeVerifierPallet::register_vk(
+                RuntimeOrigin::signed(USER_1),
+                Box::new(42)
+            ));
+            assert!(Balances::reserved_balance(USER_1) > initial_reserved_balance);
+        })
+    }
+
+    #[rstest]
+    fn fail_if_insufficient_free_balance(mut test_ext: sp_io::TestExternalities) {
+        test_ext.execute_with(|| {
+            assert_noop!(
+                FakeVerifierPallet::register_vk(RuntimeOrigin::signed(1), Box::new(42)),
+                DispatchError::Token(sp_runtime::TokenError::FundsUnavailable)
+            );
+        })
+    }
 
+    #[rstest]
+    fn not_be_allowed_for_root(mut test_ext: sp_io::TestExternalities) {
+        test_ext.execute_with(|| {
+            assert_noop!(
+                FakeVerifierPallet::register_vk(RuntimeOrigin::root(), Box::new(42)),
+                DispatchError::BadOrigin
+            );
+        })
+    }
+
+    #[rstest]
+    fn handle_double_registration_by_different_users(mut def_vk: sp_io::TestExternalities) {
+        def_vk.execute_with(|| {
+            let initial_reserved_balance = Balances::reserved_balance(USER_2);
+            assert_ok!(FakeVerifierPallet::register_vk(
+                RuntimeOrigin::signed(USER_2),
+                Box::new(REGISTERED_VK)
+            ));
+            assert!(Balances::reserved_balance(USER_2) > initial_reserved_balance);
+        })
+    }
+
+    #[rstest]
+    fn handle_double_registration_by_same_user(mut def_vk: sp_io::TestExternalities) {
+        def_vk.execute_with(|| {
+            let initial_reserved_balance = Balances::reserved_balance(USER_1);
+            assert_ok!(FakeVerifierPallet::register_vk(
+                RuntimeOrigin::signed(USER_1),
+                Box::new(REGISTERED_VK)
+            ));
+            assert_eq!(Balances::reserved_balance(USER_1), initial_reserved_balance);
+        })
```

Comment:
Also check emission of VkRegistered event ?

Reply:
Addressed in https://github.com/zkVerify/zkVerify/pull/178/commits/a4eeb34edcd2f58693e2b957c85ec532be4915e0